src/programs/RelationalPrograms.jl

```diff
@@ -145,9 +145,15 @@ function parse_relation_diagram(head::Expr, body::Expr)
     Expr(:where, expr, context) => (expr, parse_relation_context(context)...)
     _ => (head, nothing, nothing)
   end
-  var_types = if isnothing(all_types) # Untyped case.
+  var_types = if isnothing(all_types)    # Untyped case
     vars -> length(vars)
-  else # Typed case.
+  elseif typeof(all_types[1]) <: Int     # Int typed case
+    var_type_map = Dict{Symbol,Int}(zip(all_vars, all_types))
+    vars -> getindex.(Ref(var_type_map), vars)
+  elseif typeof(all_types[1]) <: Expr    # Expr typed case
```

Comment:
eltype here too. but I'm thinking you might get a Union{Symbol, Expr} or something like that.